### PR TITLE
allow defaults for cpdef functions

### PR DESCRIPTION
**Pull request by mattip**
fixes the tests added for ~gh-1834~ gh-1864, but does this break something else?

Edited to fix the issue number

### CODE REVIEW

File: Cython/Compiler/ExprNodes.py [hunk 1]
```diff
@@ -9380,7 +9380,7 @@ def analyse_default_args(self, env):
         must_use_constants = env.is_c_class_scope or (self.def_node.is_wrapper and env.is_module_scope)
 
         for arg in self.def_node.args:
-            if arg.default and not must_use_constants:
+            if arg.default:  # and not must_use_constants:
```
Comment (mattip) [inline Cython/Compiler/ExprNodes.py:9383]: If this is correct, remove the commented code

File: tests/run/py34_signature.pyx [hunk 2]
```diff
@@ -106,7 +106,7 @@ cpdef cp2(a, b=True):
     """
     >>> def py_cp2(a, b=True): pass
 
-    #>>> signatures_match(cp2, py_cp2)
+    >>> signatures_match(cp2, py_cp2)
```
Comment (mattip) [inline tests/run/py34_signature.pyx:109]: Before merging, remove the comment on line 104

**Comment by da-woods** [comment]
> fixes the tests added for [gh-1834](https://github.com/cython/cython/issues/1834), but does this break something else?

#1864 I think?

**Comment by mattip** [comment]
Many tests are failing. Is there a way to get a summary of them? Only `public_fused_types` shows up when I search for "FAIL:",  but `fused_bound_functions` is failing as well. Should the [python3.8 C](https://travis-ci.org/github/cython/cython/jobs/765836786) job pass with no errors currently on master?

**Comment by da-woods** [comment]
> Many tests are failing. Is there a way to get a summary of them? Only `public_fused_types` shows up when I search for "FAIL:", but `fused_bound_functions` is failing as well. Should the [python3.8 C](https://travis-ci.org/github/cython/cython/jobs/765836786) job pass with no errors currently on master?

I don't think there's an easy way to get a summary. I usually just search for `FAILED` (case sensitive) then scroll upwards (I think they're mostly listed with some error output directly above the "FAILED" line). I think all the CPython tests are passing on the current master (so Python3.8 C should be fine).

**Comment by scoder** [comment]
> I don't think there's an easy way to get a summary. I usually just search for FAILED

I pushed a change to master that repeats the test output of each shard after collecting all shard results. Was easier than I had thought.

**Comment by da-woods** [comment]
@mattip It might be worth checking it against https://github.com/cython/cython/pull/4118 (since I think that's playing with the same bit of code).

I think the upshot is that that bit of code in ExprNodes does two separate things:

1. Prepare the `__defaults__` and `__kwdefaults__` values for introspection (which I think is what you want to do).
2. Make Cython use an internal defaults argument on the CyFunction when calling the function.

At the moment there's no real reason for them to be linked (although one can imagine future changes where the two bits might interact...).

The other PR separates the generation of the two, which would probably be enough to make them introspectable on `cpdef` functions.

**Comment by mattip** [comment]
I am happy to defer to #4118. I will close this one.

**Pull request closed without merging**